---
title: "How to read the Methylseq report - advanced analysis"
author: "Zymo Bioinformatics"
date: "`r format(Sys.time(), '%d %B, %Y')`"
output:
  github_document:
    toc: true
    toc_depth: 3
    fig_width: 7
    fig_height: 5
    pandoc_args: --webtex=http://chart.apis.google.com/chart?cht=tx&chl=
always_allow_html: true
---

```{r setup, include=FALSE}
library(knitr)
library(kableExtra)
knitr::opts_chunk$set(echo = FALSE, warning = F,  fig.align = "center", out.width = "95%")
dataType<-"MethylSeq"
imgFolder<-file.path("../images",dataType,"Advanced")
assetFolder<-file.path("../assets", dataType)
terminologyFile<-file.path(assetFolder, "terminology.csv")
placeholder<-file.path(imgFolder,"placeholder.png")
#----------------------------------

#--------Functions----------
highlight<-function(txt, color="yellow")
{
    return(sprintf("<span style='color: %s;'>%s</span>", color,txt))
}
## a wrapper function to insert images from 'images folder'
insert_image<-function(f, ...) {
	f<-file.path(imgFolder, f)
#	cat("file:", f)
	stopifnot(file.exists(f))
	include_graphics(f, ...)
}

## a hood to add border to images
knitr::knit_hooks$set(class = function(before, options, envir) {
  if(options$class == "img-border") {
    if(before){
      sprintf("<kbd style='border: 3px solid #9ecff7;'>", options$class)
    }else{
      "</kbd>"
    }
  }
  })
```

## Overview of the pipeline

The report was generated with Zymo **Methylseq Advanced Analysis** pipeline
built on the [nextflow](https://www.nextflow.io/) platform. A example
report can be found at [here](../reports/MethylSeq_sample_report.advanced.html).

The analysis includes the following steps:

1. Detect DMCs (differential methylated cytosines); available
    statistical methods include [Student's t-test](https://en.wikipedia.org/wiki/Student%27s_t-test),
    [ANOVA](https://en.wikipedia.org/wiki/Analysis_of_variance), 
    [Fisher's exact test](https://en.wikipedia.org/wiki/Fisher%27s_exact_test),
    and [DSS](https://bioconductor.org/packages/release/bioc/html/DSS.html).

2. Detect DMRs (differential methylated regions); available
    statistical methods include [DSS](https://bioconductor.org/packages/release/bioc/html/DSS.html)
    and [dmrseq](https://bioconductor.org/packages/release/bioc/html/dmrseq.html).

3. Annotate both DMCs and DMRs by finding overlapped genes, promoters,
    exons, introns, and CpG islands.

4. Functional enrichment analysis of DMR-overlapped genes by using the
    tool [g:Profiler](https://biit.cs.ut.ee/gprofiler/gost).

5. Generate an html report with tables, figures, and downloading links.


## Report overview

This report is generated by [diffMeth](https://github.com/Zymo-Research/diffMeth).
It has a navigation box on the left and the main results on the right. And the
results are divided into sections, including workflow summary, sample information,
DMCs and DMRs, functional enrichment analysis, and file downloads.


```{r img-overview, class="img-border"}
insert_image("00.overview.png")
```

## Workflow of data processing

This section provides a summary of the data analysis, including
the thresholds used for filtering input data, the tools and parameters for analyses,
and the results generated.

```{r,  class="img-border"}
insert_image("00.workflow.png")
```

## Sample information

This section provides the sample information, including sample IDs
and their associated groups.

```{r,  class="img-border"}
insert_image("01.sample_information.png")
```

## Distribution of methylation values

This section shows the distributions of methylation values for all CpGs
using the filtered data. The distributions are made
at the both sample and group levels.

The same color scheme is used in both per-sample and group-mean distributions.

### Methylation distribution - per sample

The methylation values of all CpGs (after read depth filtering) in each sample is presented in a
[violin plot](https://en.wikipedia.org/wiki/Violin_plot). The samples are colored
to distinguish different groups.

```{r,  class="img-border"}
insert_image("02.meth_distribution.sample.png")
```

### Methylation distribution - group means

The methylation values of each CpG are averaged over samples in a group
to derive group means, which are displayed in a [density plot](https://en.wikipedia.org/wiki/Density_estimation).
Note that only CpGs that pass the filterings in all the groups are used,
and when there are more than 5 million CpGs, the sites are down-sampled to 5 million
for computing efficiency.

```{r,  class="img-border"}
insert_image("03.meth_distribution.group.png")
```


## DMCs

DMCs, short for differential methylated cytosines, can be detected by using
various kinds of statistical methods. In this section, we present a summary
of the detected DMCs and the heatmaps of top significant DMCs (reduced to
3000 if more are detected) for each group comparison.


### Summary

For each comparison, the table presents the compared groups, the number of
input sites (after filtering), the number of significant DMCs (both hyper-
and hypo-methylated) and the result's filename. The filenames can be used
to download the data from the [Downloads](#downloads) section.

```{r,  class="img-border"}
insert_image("04.DMC_summary.png")
```

### Heatmap

This section shows the heatmaps of methylation values for the DMCs detected
in each comparison. The heatmaps are organized in a slideshow and one can
click the buttons at the bottom to access different ones.

In each heatmap, the title at the top shows the groups under comparison
and the number of DMCs used (if more than 3000 DMCs detected, then only
top 3000 will be used). The dengrograms at the top and left
show the clustering of samples and sites, respectively. The clusterings
were computed using hierarchical clusterings with Eucleadian distances.

```{r,  class="img-border"}
insert_image("05.DMC_heatmap.png")
```

## DMRs

DMRs stand for differential methylated regions. They can also be detected by using
various kinds of statistical methods. This section has a similar structure as
that of [DMCs](#dmcs), so one can find a summary table as well as heatmaps
for top DMRs.


### Summary

For each comparison, the table presents the compared groups, the number of
input sites (after filtering), the number of significant DMRs (both hyper-
and hypo-methylated) and the result's filename. The filenames can be used
to download the data from the [Downloads](#downloads) section.

Note that many DMR statistical methods, such as DSS, don't compute P values,
and thus no adjusted *P* values are computed.

```{r,  class="img-border"}
insert_image("06.DMR_summary.png")
```

### Heatmap

This section shows the heatmaps of methylation values for the DMRs detected
in each comparison. Again, only top 3000 DMRs are presented. The methylation 
value of each DMR in each sample is computed as follows:

$$\frac{ \sum_{\text{sites in DMR}} {#methylated} }{ \sum_{\text{sites in DMR}}{#total} }$$

where $#total$ and $#methylated$ provide the numbers of total and methylated
reads at each site, respectively.

Similarly, the dengrograms were computed using hierarchical clusterings with Eucleadian distances.

```{r,  class="img-border"}
insert_image("07.DMR_heatmap.png")
```

### Functional enrichment

This section shows the results of functional enrichment analysis
of DMR-overlapped genes using the platform [g:Profiler](https://biit.cs.ut.ee/gprofiler/gost).
Briefly, the platform maps the provided gene list to functional terms and
finds the enriched terms. The functional terms come from various
databases (GeneOntology, KEGG, CORUM, etc). In the plot, each dot represents one functional
term with the statistical significance shown on the y-axis 
(in the format $\log_{10}(\text{adjusted P value})$).

```{r,  class="img-border"}
insert_image("08.functional_enrichment.DMR_genes.png")
```

## Downloads

This section provides links to all kinds of downloadable
results. At present, the provided are DMCs, DMRs, and functional
enrichment analysis of DMR-overlapped genes. 

```{r,  class="img-border"}
insert_image("09.downloads.png")
```

These files are normally tab- or comma- delimited, and compressed
in gzip. More details on the file formats are provided below:

### DMC files

The filename is named in the format of *dms_\<method\>.\<group1\>\_vs\_\<group2\>.tsv.gz*,
where *\<method\>* provides the statistical method (e.g., DSS), and *\<group1\>*
and *\<group2\>* are the compared group names.

Note that some columns are method-specific and thus omitted in output;
for example, the column 'postprob.overThreshold' is only available in
the output of DSS. The same principle applies to the DMR files.

```{r}
caption<-NULL
format_file<-read.csv(file.path(assetFolder,"file_format.dmc.csv"))
kable(format_file, caption=caption)
```

### DMR files

The filename is named in the format of *dmr_\<method\>.\<group1\>\_vs\_\<group2\>.tsv.gz*,
where *\<method\>* provides the statistical method (e.g., DSS), and *\<group1\>*
and *\<group2\>* are the compared group names.

The file content is explained as follows:

```{r}
caption<-NULL
format_file<-read.csv(file.path(assetFolder,"file_format.dmr.csv"))
kable(format_file, caption=caption)
```

### Functional enrichment of DMR overlapping genes files

These files are generated by [gprofiler](https://biit.cs.ut.ee/gprofiler/gost),
implemented in R package [gprofiler2](https://cran.r-project.org/web/packages/gprofiler2/index.html).
 The files show the significant functional terms. One can find more information
on the analysis and the output format by visiting [here](https://biit.cs.ut.ee/gprofiler/page/docs)
and [here](https://cran.r-project.org/web/packages/gprofiler2/vignettes/gprofiler2.html).
The functional terms examined are from multiple resources, including
[Gene Ontology](http://geneontology.org/), [KEGG](https://www.genome.jp/kegg/),
[Reactome](https://reactome.org/), [TRANSFAC](https://genexplain.com/transfac/),
[miRTarBase](https://mirtarbase.cuhk.edu.cn/~miRTarBase/miRTarBase_2019/php/index.php),
[CORUM](https://mips.helmholtz-muenchen.de/corum/),
[Human Protein Atlas](https://www.proteinatlas.org/),
[Human Phenotype Ontology](http://hpo.jax.org/app/),
and [WikiPathways](https://www.wikipathways.org/index.php/WikiPathways).


The filename is named in the format of *func_enrich.dmr_\<method\>.\<group1\>\_vs\_\<group2\>.enrich_analysis.gprofiler.tsv*,
where *\<method\>* provides the statistical method (e.g., DSS), and *\<group1\>*
and *\<group2\>* are the compared group names. Each row shows the statistical result
for each functional term.

```{r}
caption<-NULL
format_file<-read.csv(file.path(assetFolder,"file_format.gprofiler.csv"))
kable(format_file, caption=caption, escape=F, format="pipe")
```


**More on the statistical test**

Hypergeometric test is used to test whether input genes are enriched (i.e., overrepresented)
in any functional term (e.g., RNA folding, GO:0034337 from the annotation source
 GeneOntology:biological_process). Essentially, it compares the fraction of input
genes that are annotated to a functional term against the background fraction, which is
the fraction of genes annotated to the term in the entire annotation source. 

Using the figure below as an example, the rectangle denotes the entire annotation source
(e.g., GeneOntology:biological_process), and the number of genes it contains is called
*effective_domain_size*. The pink circle denotes the genes annotated to a functional
term (e.g., RNA folding, GO:0034337) in the annotation source, and the set size is called *term_size*.
The green circle denotes all the input genes, e.g., the genes overlapped with any DMRs detected
from a group comparison (genes having no annotations are excluded), and the gene set size
is called *query_size*. The overlap (blue part) between the two circles denotes the
input genes that are also annotated to the term, and its set size is called *intersection_size*.
The statistical test examines whether the fraction *intersection_size/query_size* is significantly
greater than *term_size/effective_domain_size*.


```{r, out.width="60%"}
include_graphics(file.path(assetFolder,"gprofiler-explanation.png"))
```
