---
title: "How to read the report of RRBS basic analysis"
author: "Zhenguo Zhang"
date: "`r format(Sys.time(), '%d %B, %Y')`"
output:
  github_document:
    toc: true
    toc_depth: 3
    fig_width: 7
    fig_height: 5
always_allow_html: true
---

```{r setup, include=FALSE}
library(knitr)
library(kableExtra)
knitr::opts_chunk$set(echo = FALSE, warning = F,  fig.align = "center", out.width = "100%")
dataType<-"MethylSeq"
imgFolder<-file.path("../images",dataType,"Basic")
assetFolder<-file.path("../assets", dataType)
terminologyFile<-file.path(assetFolder, "terminology.csv")
placeholder<-file.path(imgFolder,"placeholder.png")
#----------------------------------

#--------Functions----------
highlight<-function(txt, color="yellow")
{
    return(sprintf("<span style='color: %s;'>%s</span>", color,txt))
}
## a wrapper function to insert images from 'images folder'
insert_image<-function(f, ...) {
	f<-file.path(imgFolder, f)
	cat("file:", f)
	stopifnot(file.exists(f))
	include_graphics(f, ...)
}
```

## Overview of the pipeline

The report was generated with our **RRBS Basic Analysis** pipeline
built on [nextflow](https://www.nextflow.io/) platform. A example
report can be found at [here](../reports/MethylSeq_sample_report.basic.html).

The backbone of the pipeline is as follows:

1. Trimming input sequencing reads using
   [Trim Galore](https://github.com/FelixKrueger/TrimGalore).

2. Evaluating reads quality using
   [FastQC](https://github.com/s-andrews/FastQC).

3. Aligning trimmed reads using
   [bismark](https://github.com/FelixKrueger/Bismark).

4. Calling methylation levels using
   [MethylDackel](https://github.com/dpryan79/MethylDackel).

5. Calculating read depths per cytosine genome-wide as well as in
   different genomic regions (gene body, promoter, etc) using
   [bedtools](https://bedtools.readthedocs.io/en/latest/).

6. Getting the distribution of library insert size using
   [picard](https://gatk.broadinstitute.org/hc/en-us/articles/360037055772-CollectInsertSizeMetrics-Picard-).

7. Calculating library quality metrics by analyzing the data of
   **Zymo *In situ* Spike-in Control**.

8. Generating downloadable files (fastq, bam, bedgraph,
   UCSC track, and figures).

## Terminology

```{r load-terminology}
cat(getwd())
dat<-read.csv(terminologyFile)
kbl(dat) %>%
  kable_styling(
    bootstrap_options=c("striped", "hover"),
    full_width=F)
```

## Report overview

The bioinformatics report is generated using [MultiQC](https://multiqc.info/), which
is integrated into the nextflow pipeline. There are
general instructions on how to read a MultiQC report at 
[here](https://multiqc.info/docs/#using-multiqc-reports), or you
can watch [this video](https://www.youtube.com/watch?v=qPbIlO_KWN0).
 In general, the report has a
navigation bar to the left, allowing to quickly navigate
to any section in the report. On the right side, there is
a toolbox that allows to customize the appearance of the report
and export figures and data. Most sections of the report are
interactive: sample names and data values show up
when you mouse over them.

```{r img-overview}
insert_image("overview.png")
```

## General Statistics

The table *General Statistics* gives you an overview of the number
of aligned reads and the number of captured CpGs. Here
are some columns you can use to assess the data:

* __% Uniquely Aligned__ This is the percentage of total uniquely
  aligned read
  pairs (or reads in single-end sequencing)
  to the target genome. The total input reads is shown in column `# Reads`. In general, the higher
  the percentage, the better. The value varies over sample
  types and qualities. Normally, a value over `r highlight("50%")`
  is expected.

* __% BS Conversion (Spike-in)__ This is the bilsulfite conversion
  rate computed based on the reads aligned to our **Zymo *In situ* Spike-in Control**,
  which was added to samples in sequencing library prepration.
  A value higher than 99% indicates successful conversion
  process. A related column in the table is
  `% BS Conversion (Non-CpG)`, which is computed based on the
  reads aligned to the genome and assuming that cytosines in
  CHG and CHH contexts are not methylated; the assumption may
  not always be true (like plants) and thus this value essentially represents
  the nonmethylation levels of non-CpG cytosines.

* __Uniq. CpG__ This is the number of cytosines in CpG context
  in the genome captured by the sequenced reads. For RRBS,
  this number is expected to be smaller than WGBS, because the
  reads are derived from a portion of genome.

* __Avg. CpG Coverage__ This is the average number of reads
  covering each cytosine as reported in the `Uniq. CpG` column.
  This number is expected to be around 10 or higher for an RRBS
  library. Please also refer to the section [CpG Coverage] for
  the counts of cytosines at different read coverage level.

## Cutadapt

TrimGalore internally called cutadapt to trim low-quality bases
and adapters from reads, and any reads shorter than a 
pre-specified length (default 20) are filtered out. In this 
section, we present the counts of reads that pass filtering
and the distribution of the lengths of trimmed fragments.

### Filtered Reads

In this subsection, it shows the number of reads that passes the 
filtering of TrimGalore. TrimGalore trims low-quality bases and
adapters from the 3'end of each read and discards reads that
are shorter than 20bps. And in the paired-end mode, both reads
in a pair are discarded if at least one read is shorter than
20 bps, and this is why the read1 and read2 files of one sample
always have the same number of filtered reads.

For RRBS sequencing, the option `--rrbs` for TrimGalore is on
to remove filled-in bases at read ends, which result from
the digestion process when using the restriction enzyme MspI.
For more details on the base removal, please refer to the
[TrimGalore Manual](https://github.com/FelixKrueger/TrimGalore/blob/master/Docs/Trim_Galore_User_Guide.md#step-2-adapter-trimming).

One can toggle the tabs between *Counts* and *Percentages*
to view the numbers and percentages of filtered reads, a
feature available for most plots in the report.

### Trimmed Sequence Lengths

This plot shows the distributions of the lengths of trimmed
segments from reads (the part that is removed from a read).
In general, you will see the frequency drop quickly with
the increase of trimmed segments' lengths, meaning that
only a few bases are trimmed for most reads. For RRBS, however
you may see some small peaks in the middle; this can be caused
by small insert sizes of many fragments in a library. 

The tab `Obs/Exp` presents the ratio of observed and expected
counts for a given trimmed length. The expected count is computed
by assuming sequencing error only. A ratio higher than 1 indicates
that some trimmed segments are true adapters. One can see 
[cutadapt's guide](https://cutadapt.readthedocs.io/en/latest/guide.html#how-to-read-the-report) for more explanation.

## FastQC (trimmed)

This section shows the analysises by *FastQC* ran on trimmed
fastq files. [FastQC](http://www.bioinformatics.babraham.ac.uk/projects/fastqc/Help/)
is a tool to analyze library qualities by examining the metrics
such as base qualities, GC content, overrepresented and adapter
sequences. A warning is issued if any metric fails.

### Sequence Quality Histograms

This section shows the average [Phred quality scores](https://gatk.broadinstitute.org/hc/en-us/articles/360035531872-Phred-scaled-quality-scores)
per base along read length. Normally, a decline of scores
over read length is expected, because the base quality decreases
towards 3'end. This provides information on whether 3'end quality
trimming is needed.

### Per Sequence Quality Scores

This plot shows the distributions of read sequence quality which
is computed by averaging the Phred scores of all bases in a read.
It is expected that peaks are at values $>$ 28; if you see peaks
at lower values, it is a warning of low quality libraries.

### Per Base Sequence Content

This plot shows the percentages of the four nucleotides 
(A, T, C, G) along read length. The heatmap is presented
with samples along rows and positions along columns, and
the color represents the most frequent base at a position.
When hovering over the plot, one will see the nucleotide
compositions at the top.

One can click on one row to have clearer view on how nucleotide
composition changes over read length. The composition is
expected to be even over read length. The sharp drop of 'A' 
frequency at the last position is caused by stringent adapter
trimming, because 'A' matches the first base of Illumina
adapters.

### Per Sequence GC Content

This plot shows the distributions of reads' GC content, that
is, the percentages of G and C nucleotides in a read. 

### Overrepresented sequences

If there are overrepresented sequences, such as contamination,
this section will show these sequences. A good library should
have no such sequences.

### Adapter Content

This plot shows the percentages of reads having an adapter
sequence at each position along read length. A good
library should see zero percentage along read length until
3' end where adapter sequences are expected. If there are
many adapters present in the middle, a further diagnosis is
needed.

## Bismark

[Bismark](https://www.bioinformatics.babraham.ac.uk/projects/bismark/)
is a tool to align bisulfite-converted sequencing reads to genome.
One can find the manual of the program at [here](https://github.com/FelixKrueger/Bismark).

### Alignment Rates

This plot shows the number and percentage of reads falling into
the following categories:

* Aligned Uniquely: reads that have a unique mapping position in the target genome.

* Aligned Ambiguously: reads that are alignable to genome, but mappable to multiple genomic positions.

* Did Not Align: reads not alignable to genome.

For downstream analyses such as calling methylation, only *Aligned Uniquely*
reads are used.

### Strand Alignment

This plot shows which strand each read pair is aligned. Due to
bisulfite conversion, there are four strands that a read pair
can align:

1. Original top strand: reads derived from top/Waston strand.

2. Complementary to original top strand: reads derived from top/Waston strand but in a complementary form, generated by PCR.

3. Original bottom strand: reads derived from bottom/Crick strand.

4. Complementary to original bottom strand: reads derived from bottom/Crick strand but in a complementary form, generated by PCR.

For a directional sequencing library, you may only see reads
from case 1 and 3, but for non-directional one, you expect a
even distribution among the four strands.

### M-bias

This plot presents the average methylation value over all reads
along read positions. Normally, one expect the value stay the same
along read positions.  

## Insert Size

This shows the distribution of estimated insert sizes for each
sample. Usually, one expect the size clusters around a certain
value, determined by size selection in library preparation. For
RRBS, due to the nature of MspI cut, one may see multiple peaks
in the range 40 to 220 bps (see explanation [here](http://www.bioinformatics.babraham.ac.uk/projects/bismark/RRBS_Guide.pdf)).

## CpG Coverage

This plot presents the number and percentage of cytosines in CpG
context under different read coverages. Here, only cytosines
covered by at least one read are considered.
For better visualization,
the read coverage is grouped into four categories: 1-4, 5-9, 
10-49, >=50. For RRBS, one may see many cytosines in the 
category 10-49. If you see many cytosines in the lower
categories, a higher sequencing depth may be needed if
you want to investigate those low-coverage cytosines.

## Genomic Region Coverage

This section presents how well functional regions
are represented in the libraries. Here, the functional regions
include CpG island, gene body, and promoter. For a functional
region, following the procedure in `r highlight("citation")`,
we define the coverage as the number of reads covering any
cytosine in that region.

### CpG Island

CpG islands are clusters of CpGs and defined operatively.
Here, we use the definition by 
[Gardiner-Garden M and Frommer M.](https://www.ncbi.nlm.nih.gov/pubmed/3656447),
and use the max scoring algorithm to identify them. More on the
CpG island criteria can be found at [UCSC](https://genome.ucsc.edu/cgi-bin/hgTrackUi?hgsid=998595637_fiKBxHtrl6naFyjEJbvl4P9Nhuc0&c=chr12&g=cpgIslandExt).

As you can see, CpG islands are divieded into groups based on
their coverage. You can toggle between `Counts` and `Percentages`.

### Gene Body

A similar analysis to the CpG island. Here gene body is
defined as the region between a gene's first and last base
in a genome, so including both exons and introns. We use
the gene annotation from `r highlight("missing")`.

### Promoter

Promoter is defined as the 2000 bps immediately before the
start of a gene. We use
the gene annotation from `r highlight("missing")`.

## Bismark (spike-in)

This section shows the numbers of reads aligned to all spike-in
sequences for each sample. Similar to the presentation for reads aligned to genome,
we also show the number of alignments on each strand.

## Samtools (spike-in)

Similar to the above, but the reads are divided based on which
spikein sequence (amplicon) they are aligned. There are totally
6 amplicons.

## MethylDackel (spikein)

In this section, we show the correlation between observed
and expected methylation values for the spikein sequences.
The methylation values are extracted from bam files using
[MethylDackel](https://github.com/dpryan79/MethylDackel).
Since the methylation value for each spikein sequence is
known, this analysis helps to evaluate the library qualities.
For example, a significant deviation of observed methylations
from expected ones can be a sign of incomplete bisulfite
conversion among other issues.

### Correlation Scatter Plot

This subsection presents a correlation plot of expected (x axis)
and observed (y axis) methylation values for all spikein
sequences. The points are expected to fall around a slope line of 
45 degrees.

### Correlation Table

This table presents the `r highlight("Pearson", "red")`
correlation coefficient for the  expected 
and observed methylation values of spikein sequences,
which is a more quantitative measurement than the above
correlation plots.

## Download

This section provides links to download all kinds of
data the pipeline generated.

Note that the data are available for download in a limited time,
after which, they will be taken down.

### Methylation Distribution

These are figures containing methylation value distributions
in functional regions for each sample. The functional regions
include CpG islands, gene bodies, and promoters. With the
violin plots, one can see how methylation differ among the
functional regions.

### Bam

The bam files generated by bismark.

### Browser Track

The UCSC browser tracks in [bigbed format](https://genome.ucsc.edu/goldenPath/help/bigBed.html). One can view these
files directly in the UCSC browser. An instruction on how
to view bigbed files in UCSC custom tracks can be found
[here](https://genome.ucsc.edu/goldenPath/help/bigBed.html).

### Trimmed Fastq R2

These are Read 2 fastq files after trimming by TrimGalore.

### Trimmed Fastq R1

These are Read 1 fastq files after trimming by TrimGalore.

### Bedgraph

These are bedgraph files containing the numbers of reads
supporting methylation and nonmethylation statuses for
each cytosine in a genome. These files are generated with
[MethylDackel](https://github.com/dpryan79/MethylDackel).

Each file has 6 columns (separated by tab) as follows:

1. The chromosome name.
2. The start coordinate of a cytosine on the chromosome.
3. The end coordinate of a cytosine on the chromosome.
4. The methylation percentage rounded to an integer.
5. The number of reads/pairs reporting methylation status.
6. The number of reads/pairs reporting nonmethylation status.

One can find more information on the format by checking
MethylDackel page.

